### ISSUE TITLE
Native menu enable/disable management based on project selection

### ISSUE THREAD

**Issue by rat-moonshine** [prominic/Moonshine-IDE#274]
Menu enable/disable problem come up again when #265 tests; As reported scenario by @JoelProminic , 

> Open Visual Editor Project
> File > Recent Project >
> Close the additional project. The VisualEditor project is now selected
> File > Recent Project. BUG: menu items are disabled

This is possibly because the separate collection management maintains for visual editor menu enable/disable purpose. 

At this moment to enable/disable menu between Visual Editor and non/Visual-Editor projects, it maintain manually separately managed collections like this:

```
public static var menuItemsEnabledInVEProject:Array = [
            resourceManager.getString('resources', 'NEW'),
            resourceManager.getString('resources', 'OPEN'),
			resourceManager.getString('resources', 'OPEN_RECENT_PROJECTS'),
			resourceManager.getString('resources', 'OPEN_RECENT_FILES'),
.....
.....
            resourceManager.getString('resources', 'RENAME'),
            resourceManager.getString('resources', 'VISUALEDITOR_FLEX_FILE'),
            resourceManager.getString('resources', 'VISUALEDITOR_PRIMEFACES_FILE'),
            "Copy Path",
            "Show in Explorer",
            "Show in Finder"
        ];

        private static var menuItemsDisabledNoneVEProject:Array = [
            resourceManager.getString('resources', 'EXPORT_VISUALEDITOR_PROJECT'),
            resourceManager.getString('resources', 'VISUALEDITOR_FLEX_FILE'),
            resourceManager.getString('resources', 'VISUALEDITOR_PRIMEFACES_FILE'),
            resourceManager.getString('resources', 'EXPORT_VISUALEDITOR_PROJECT_TO_FLEX'),
            resourceManager.getString('resources', 'EXPORT_VISUALEDITOR_PROJECT_TO_PRIMEFACES')
        ];
```

I don't like this design, and I think this can be improve. The current way is unexpected-scenario-prone because of the following reasons:

1. Short manually managed collection is not problem unless it grows larger
2. Forget to add any newly added native menu to this list cause to be disable 
3. Unexpected results with dynamic populated menus like one reported by Joel on #265 

We need the enable/disable management to be more robust that the API serve almost all the project types irrespective of Visual Editor project only. I also don't want to manage different likewise (above) menu collection set for each of the project types those supports by Moonshine, to enable/disable menus between projects. We need to make the process more generic than specific.

**Comment by piotrzarzycki21** [comment]
Agree with you! I don't like it either, but when you go deeper into all of that code you will understand that we have a bit of a mess there. That menu is created separately for Mac and Windows, which should not have place, even if it's looks different. In my opinion whole Menu plugin should be removed and written from scratch. We should use XML to describe menu instead above idea. It's definitely work not for couple of hours, but for days. 

My 2 cents. :)

**Comment by rat-moonshine** [comment]
By OS structure, and the behavior (accompanied by the OS), I don't find any not-good thing using different types of menus based on the platform. Of course, you could also use same custom menu that used in Windows to macOS - but macOS native menu definitely feels more organic and native. But we can't use that macOS native menu in Windows.

On a second thought, the author could have used the Flex Menu API instead of the custom menu (that is using now). But I think he might thought about having more custom properties to be available and other things; and, this is stable anyway.

Default menu generation can be populated by an XML, but this is not a priority thing.

**Comment by rat-moonshine** [comment]
Okay, I have done this requirement changes. From now on, declaration of a `MenuItem` also requires to pass a new props `enableTypes` to avail the menu option to the particular choices (types) of projects that passed by `enableTypes`.

The updated `MenuItem` declaration is now looks like this:

`public function MenuItem(label:String , items:Array=null, enableTypes:Array=null, ....`

The `enableTypes` can pass with an array of String types those declared into a new class file `ProjectMenuTypes`, or `null`:

- `null` : Avail the menu item to every project irrespective of the project type
- `[ProjectMenuTypes.FLEX_AS, ProjectMenuTypes.JS_ROYALE]` : this will avail the menu option only to the project that has a type `FLEX_AS` or `JS_ROYALE`
- A new `menuType` introduced to the `AS3ProjectVO` object which shall hold the project menu type from beginning (when a project loads or creates), default is `FLEX_AS`
- `MenuPlugin` will continue enable/disable menu items based on the menu item's `enableTypes` and active project's `menuType`

So menu declaration in `IFlexCoreBridgeImp` file now looks like this (as an example a menu that only available to Visual Editor PrimeFaces project):

`new MenuItem(resourceManager.getString('resources', 'EXPORT_VISUALEDITOR_PROJECT_TO_PRIMEFACES'), null, [ProjectMenuTypes.VISUAL_EDITOR_PRIMEFACES], ExportVisualEditorProjectEvent.EVENT_EXPORT_VISUALEDITOR_PROJECT_TO_PRIMEFACES)`

Available project menu types are as following, new project types can get added to the list:

- `FLEX_AS`: ActionScript projects, Flex Desktop/Mobile projects, Flex Mobile projects, Away3D projects, Feathers Desktop/Mobile projects
- `JS_ROYALE`: Royale/FlexJS Browser projects
- `VISUAL_EDITOR_FLEX`: Visual Editor Flex projects
- `VISUAL_EDITOR_PRIMEFACES`: Visual Editor PrimeFaces projects
- `LIBRARY_FLEX_AS`: Library projects

Give this a try, and report if any error/problem.

**Comment by piotrzarzycki21** [comment]
I would propose one change. 

I have to admit that I don't like that we have another property in `AS3ProjectVO `, but don't have an idea how to do that differently. However I would get rid off method `getProjectMenuType` from `CreateProject` and simply move that logic inside `AS3ProjectVO`. All the properties which are needed to specify what kind of menu we have are inside that object. 

What do you think ?

**Comment by rat-moonshine** [comment]
I'm trying to understand what would be the plus point to move the `getProjectMenuType` to AS3ProjectVO? The method calls only once during the project creation to set the project's `menuType` (and its essential to store the individual menu type to individual AS3ProjectVO rather that in different location, in my understanding). In that respect, `getProjectMenuType` do not serve any further purpose that it requires to be a part of AS3ProjectVO. 

Can you enlighten (?)

**Comment by piotrzarzycki21** [comment]
My idea is to have `menuType ` with all of that condition inside `AS3ProjectVO`. I'm thinking about scenario where for whatever reason `AS3ProjectVO` changed type somewhere on the fly - than that property won't work correctly. 

However I'm not forcing you to change that - if you wanted to leave it as is let it be.

**Comment by rat-moonshine** [comment]
No, thank you for proposing your thoughts, you're always welcome to do that. There are few reasons that I don't want a conditional set of method to be a part of the AS3ProjectVO, though:

1. We never had any scenario where we would want to change project's type on fly (or if we even want to do that)
2. I don't want to process same conditional checks every time against each menu item; A single field check is much more process-saving
3. In current scenario the method run only once and set a property which serve step. 2

But again, thank you for proposing with thoughts target for improving! Please, let me know if you would found any error/enable-disable problem during your tests, too.

**Comment by piotrzarzycki21** [comment]
I didn't notice any problems form yesterday, so you can close that one.
